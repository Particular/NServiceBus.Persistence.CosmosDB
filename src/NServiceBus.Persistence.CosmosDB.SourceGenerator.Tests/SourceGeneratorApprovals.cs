namespace NServiceBus.AzureFunctions.SourceGenerator.Tests
{
    using System;
    using System.Collections.Generic;
    using System.Collections.Immutable;
    using System.Linq;
    using Microsoft.CodeAnalysis;
    using Microsoft.CodeAnalysis.CSharp;
    using NUnit.Framework;
    using Particular.Approvals;
    using Persistence.CosmosDB.SourceGenerator;

    [TestFixture]
    public class SourceGeneratorApprovals
    {
        [Test]
        public void UsingNamespace()
        {
            var source =
@"using NServiceBus;

namespace Foo
{
    public class Startup
    {
        public const string EndpointName = ""endpoint"";
    }
}";
            var (output, _) = GetGeneratedOutput(source);

            Approver.Verify(output);
        }

        [Test]
        public void UsingFullyQualifiedAttributeName()
        {
            var source =
@"

namespace Foo
{
    public class Startup
    {
        public const string EndpointName = ""endpoint"";
    }
}";
            var (output, _) = GetGeneratedOutput(source);

            Approver.Verify(output);
        }

        [OneTimeSetUp]
        public void Init()
        {
        }

        static (string output, ImmutableArray<Diagnostic> diagnostics) GetGeneratedOutput(string source, bool suppressGeneratedDiagnosticsErrors = false)
        {
            var syntaxTree = CSharpSyntaxTree.ParseText(source);
            var references = new List<MetadataReference>();
            var assemblies = AppDomain.CurrentDomain.GetAssemblies();

            foreach (var assembly in assemblies)
            {
                if (!assembly.IsDynamic && !string.IsNullOrWhiteSpace(assembly.Location))
                {
                    references.Add(MetadataReference.CreateFromFile(assembly.Location));
                }
            }

            var compilation = Compile(new[]
            {
                syntaxTree
            }, references);

            var generator = new PartitionKeyMappingGenerator();

            var driver = CSharpGeneratorDriver.Create(generator);
            driver.RunGeneratorsAndUpdateCompilation(compilation, out var outputCompilation, out var generateDiagnostics);

            // add necessary references for the generated trigger
            // references.Add(MetadataReference.CreateFromFile(typeof(ServiceBusTriggerAttribute).Assembly.Location));
            // references.Add(MetadataReference.CreateFromFile(typeof(ExecutionContext).Assembly.Location));
            // references.Add(MetadataReference.CreateFromFile(typeof(Message).Assembly.Location));
            // references.Add(MetadataReference.CreateFromFile(typeof(ILogger).Assembly.Location));
            Compile(outputCompilation.SyntaxTrees, references);

            if (!suppressGeneratedDiagnosticsErrors)
            {
                Assert.False(generateDiagnostics.Any(d => d.Severity == DiagnosticSeverity.Error), "Failed: " + generateDiagnostics.FirstOrDefault()?.GetMessage());
            }

            return (outputCompilation.SyntaxTrees.Last().ToString(), generateDiagnostics);
        }

        static CSharpCompilation Compile(IEnumerable<SyntaxTree> syntaxTrees, IEnumerable<MetadataReference> references)
        {
            var compilation = CSharpCompilation.Create("result", syntaxTrees, references, new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary));

            // Verify the code compiled:
            var compilationErrors = compilation
                .GetDiagnostics()
                .Where(d => d.Severity >= DiagnosticSeverity.Warning);
            Assert.IsEmpty(compilationErrors, compilationErrors.FirstOrDefault()?.GetMessage());

            return compilation;
        }
    }
}