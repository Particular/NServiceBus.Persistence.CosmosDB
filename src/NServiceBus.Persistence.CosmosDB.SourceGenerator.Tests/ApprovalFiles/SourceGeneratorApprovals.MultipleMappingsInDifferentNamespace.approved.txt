// <auto-generated/>
using System;
using Microsoft.Azure.Cosmos;
using NServiceBus.Persistence.CosmosDB;
using Some.Other.Namespace;
using Yet.Another.Namespace;
using System.Collections.Generic;
using NServiceBus;

namespace Some.Complex
{
    internal partial class PartitionKeyMapper
    {

        protected override bool TryMapMessageCore(object message, out PartitionKey? partitionKey, out ContainerInformation? containerInformation)
        {
            partitionKey = null;
            containerInformation = null;
            if(message is IProvideOrderId)
            {
                var (invoker, container, state) = Extractors[typeof(IProvideOrderId)];
                partitionKey = invoker(message, state);
                containerInformation = container;
                return true;
            }
            if(message is OrderDeclined)
            {
                var (invoker, container, state) = Extractors[typeof(OrderDeclined)];
                partitionKey = invoker(message, state);
                containerInformation = container;
                return true;
            }
            return false;
        }

    }
}